src/quacc/recipes/espresso/_base.py

```diff
@@ -92,3 +92,84 @@ def base_fn(
     return summarize_run(
         final_atoms, input_atoms=atoms, additional_fields=additional_fields
     )
+
+
+def base_opt_fn(
+    atoms: Atoms = None,
+    preset: str | None = None,
+    relax_cell: bool = False,
+    template: EspressoTemplate | None = None,
+    profile: EspressoProfile | None = None,
+    calc_defaults: dict[str, Any] | None = None,
+    calc_swaps: dict[str, Any] | None = None,
+    opt_defaults: dict[str, Any] | None = None,
+    opt_params: dict[str, Any] | None = None,
+    parallel_info: dict[str, Any] | None = None,
+    additional_fields: dict[str, Any] | None = None,
+    copy_files: list[str] | None = None,
+) -> RunSchema:
+    """
+    Base function to carry out espresso recipes.
+
+    Parameters
+    ----------
+    atoms
+        Atoms object
+    preset
+        Name of the preset to use
+    relax_cell
+        Whether to relax the cell or not.
+    template
+        EspressoTemplate to use
+    profile
+        EspressoProfile to use
+    calc_defaults
+        The default calculator parameters.
+    calc_swaps
+        Custom kwargs for the espresso calculator. Set a value to
+        `quacc.Remove` to remove a pre-existing key entirely. For a list of available
+        keys, refer to the `ase.calculators.espresso.Espresso` calculator.
+    opt_defaults
+        The default optimization parameters.
+    opt_params
+        Dictionary of parameters to pass to the optimizer. pass "optimizer"
+        to change the optimizer being used. "fmax" and "max_steps" are commonly
+        used keywords. See the ASE documentation for more information.
+    parallel_info
+        Dictionary of parallelization information.
+    additional_fields
+        Any additional fields to supply to the summarizer.
+    copy_files
+        Files to copy to the runtime directory.
+
+    Returns
+    -------
+    RunSchema
+        Dictionary of results from [quacc.schemas.ase.summarize_run][]
+    """
+
+    atoms = Atoms() if atoms is None else atoms
+
+    calc_defaults["input_data"] = Namelist(calc_defaults.get("input_data"))
+    calc_swaps["input_data"] = Namelist(calc_swaps.get("input_data"))
+
+    binary = template.binary if template else "pw"
+
+    calc_defaults["input_data"].to_nested(binary=binary, **calc_defaults)
+    calc_swaps["input_data"].to_nested(binary=binary, **calc_swaps)
+
+    calc_flags = recursive_dict_merge(calc_defaults, calc_swaps)
```

Comment:
There is some content here copied and pasted between `base_fn` and `base_opt_fn`. It would be good to consolidate that, if practical. The easiest option is probably a custom function called by both.